Resolves #267: New fragment MultiSelect

Baret opened the pull request:
Analogous to Components the builder for MultiSelect is provided by the new object Fragments. More fragments might be added the same way.

Baret commented:
Here is a screen capture of the example application (`MultiSelectExample`):
![multiselect_example_application](https://user-images.githubusercontent.com/450955/72665326-8e038980-3a07-11ea-8958-0823b008e84a.gif)

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/builder/fragment/MultiSelectBuilder.kt:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * The callback to be used when the value changes. It gets the old and the new value as parameters.
+     */
+    fun withCallback(callbackFunction: (oldValue: T, newValue: T) -> Unit) = also {
```
This is going to be obsolete soon when `ListProperty` gets implemented in Cobalt. Let's keep it here until then, but keep it in mind for the future.

Baret replied:
Oh, you are right. The whole idea of this fragment is a prime example for a property to watch / be watched (don't know why that wasn't my first idea, but it would have collided with the old/new value thingy anyways). Looking forward for the ListProperty.

adam-arold replied:
Btw you can use `Property<PersistentList<T>>` if you really want to, but it would be a bit awkward. 😆

adam-arold replied:
It just occurred to me that you can have a `selectedValue: Property<T>` here and then you don't need the callback at all.

Baret replied:
And when the label is clicked the property updates with the same value, right? I think that would be much easier. But on the other hand you lose the ability to detect if the value changed or not (which I for one wouldn'nt need).

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/Fragments.kt:
```diff
@@ -0,0 +1,15 @@
+import kotlin.jvm.JvmStatic
+
+/**
+ * Provides builders for [Fragment]s (more complex Components that come with own logic).
```
```suggestion
 * Provides builders for [Fragment]s (more complex Components that come with their own logic).
```

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/builder/fragment/MultiSelectBuilder.kt:
```diff
@@ -0,0 +1,70 @@
+         * @param values The values to cycle through
+         * @param N The type of the values
+         */
+        fun <N: Any> newBuilder(width: Int, values: List<N>) = MultiSelectBuilder(width, values)
```
There is a missing `@JvmStatic` here.

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/fragment/MultiSelect.kt:
```diff
@@ -0,0 +1,18 @@
+ */
+interface MultiSelect<T: Any>: Fragment {
+    val values: List<T>
+    val callback: (oldValue: T, newValue: T) -> Unit
```
You don't need this as part of the interface as the callback is only used / added in the builder.

Baret replied:
I was thinking quite hard about what to put into the interface as the documentation describes rather precisely how a `MultiSelect` should behave but the interface did not offer any methods or properties to force implementations what to use. In the first attempt it was simply empty, so I added some basic stuff (more or less at radom). `width` almost made it, too :D
I will remove callback though.

adam-arold replied:
I only meant `callback`, `values` is fine there.

Baret replied:
👍

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/internal/fragment/impl/DefaultMultiSelect.kt:
```diff
@@ -0,0 +1,119 @@
+
+    private fun getStringValue(index: Int) = toStringMethod.invoke(values[index]).centered()
+
+    private fun String.centered(): String {
```
I really like this extension function! 👍

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/internal/fragment/impl/DefaultMultiSelect.kt:
```diff
@@ -0,0 +1,119 @@
+        override val callback: (oldValue: T, newValue: T) -> Unit,
+        private val centeredText: Boolean = true,
+        private val toStringMethod: (T) -> String = Any::toString,
+        private val clickable: Boolean = false
```
This is only used when initializing this object so it doesn't have to be a field.

Baret replied:
That's absolutely correct, but why did IntelliJ not tell me? =(

adam-arold commented on zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/builder/fragment/MultiSelectBuilder.kt:
```diff
@@ -0,0 +1,70 @@
+        this.clickable = clickable
+    }
+
+    override fun build(): MultiSelect<T> = DefaultMultiSelect(width, values, callback, centeredText, toStringMethod, clickable)
```
Since this constructor has a very long parameter list I'd consider using keyword arguments here. This makes the code more resilient to future changes and it can avoid some nasty bugs when there are multiple positional parameters with the same type.

Baret commented:
Updated and changed most parts you mentioned. I am still thinking about using a `Property<T>` instead of the callback method. It feels like the callback is more flexible, but that isn't true, right? A Property has the same effect (calling some code when the value changes), except you do not know what the old value was. And of course it would be the "Cobalt way of lilfe" ;)

adam-arold commented:
It is just that we're using properties in all other components. The reason why a `Property` is useful here is that you not only have the option to listen to change events but you can also bind properties together. For example here

```kotlin
val prop0 = 1.toProperty()
val prop1 = 2.toProperty()
val binding = prop0 + prop1
```

we create a binding which will have the value of `3` and it will be updated whenever `prop0` or `prop1` changes.
